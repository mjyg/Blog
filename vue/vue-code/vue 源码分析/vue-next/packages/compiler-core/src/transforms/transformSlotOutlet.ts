import { NodeTransform } from '../transform'
import {
  NodeTypes,
  CallExpression,
  createCallExpression,
  ExpressionNode
} from '../ast'
import { isSlotOutlet } from '../utils'
import { buildProps } from './transformElement'
import { createCompilerError, ErrorCodes } from '../errors'
import { RENDER_SLOT } from '../runtimeHelpers'

export const transformSlotOutlet: NodeTransform = (node, context) => {
  if (isSlotOutlet(node)) {
    const { props, children, loc } = node
    const $slots = context.prefixIdentifiers ? `_ctx.$slots` : `$slots`
    let slotName: string | ExpressionNode = `"default"`

    // check for <slot name="xxx" OR :name="xxx" />
    let nameIndex: number = -1
    for (let i = 0; i < props.length; i++) {
      const prop = props[i]
      if (prop.type === NodeTypes.ATTRIBUTE) {
        if (prop.name === `name` && prop.value) {
          // static name="xxx"
          slotName = JSON.stringify(prop.value.content)
          nameIndex = i
          break
        }
      } else if (prop.name === `bind`) {
        const { arg, exp } = prop
        if (
          arg &&
          exp &&
          arg.type === NodeTypes.SIMPLE_EXPRESSION &&
          arg.isStatic &&
          arg.content === `name`
        ) {
          // dynamic :name="xxx"
          slotName = exp
          nameIndex = i
          break
        }
      }
    }

    const slotArgs: CallExpression['arguments'] = [$slots, slotName]
    const propsWithoutName =
      nameIndex > -1
        ? props.slice(0, nameIndex).concat(props.slice(nameIndex + 1))
        : props
    let hasProps = propsWithoutName.length > 0
    if (hasProps) {
      const { props: propsExpression, directives } = buildProps(
        node,
        context,
        propsWithoutName
      )
      if (directives.length) {
        context.onError(
          createCompilerError(
            ErrorCodes.X_V_SLOT_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
            directives[0].loc
          )
        )
      }
      if (propsExpression) {
        slotArgs.push(propsExpression)
      } else {
        hasProps = false
      }
    }

    if (children.length) {
      if (!hasProps) {
        slotArgs.push(`{}`)
      }
      slotArgs.push(children)
    }

    node.codegenNode = createCallExpression(
      context.helper(RENDER_SLOT),
      slotArgs,
      loc
    )
  }
}
