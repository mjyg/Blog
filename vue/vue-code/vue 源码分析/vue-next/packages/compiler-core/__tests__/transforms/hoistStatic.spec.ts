import {
  parse,
  transform,
  NodeTypes,
  generate,
  CompilerOptions
} from '../../src'
import {
  OPEN_BLOCK,
  CREATE_BLOCK,
  CREATE_VNODE,
  WITH_DIRECTIVES,
  FRAGMENT,
  RENDER_LIST
} from '../../src/runtimeHelpers'
import { transformElement } from '../../src/transforms/transformElement'
import { transformExpression } from '../../src/transforms/transformExpression'
import { transformIf } from '../../src/transforms/vIf'
import { transformFor } from '../../src/transforms/vFor'
import { transformBind } from '../../src/transforms/vBind'
import { transformOn } from '../../src/transforms/vOn'
import { createObjectMatcher, genFlagText } from '../testUtils'
import { PatchFlags } from '@vue/shared'

function transformWithHoist(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)
  transform(ast, {
    hoistStatic: true,
    nodeTransforms: [
      transformIf,
      transformFor,
      ...(options.prefixIdentifiers ? [transformExpression] : []),
      transformElement
    ],
    directiveTransforms: {
      on: transformOn,
      bind: transformBind
    },
    ...options
  })
  expect(ast.codegenNode).toMatchObject({
    type: NodeTypes.JS_SEQUENCE_EXPRESSION,
    expressions: [
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: OPEN_BLOCK
      },
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_BLOCK
      }
    ]
  })
  return {
    root: ast,
    args: (ast.codegenNode as any).expressions[1].arguments
  }
}

describe('compiler: hoistStatic transform', () => {
  test('should NOT hoist root node', () => {
    // if the whole tree is static, the root still needs to be a block
    // so that it's patched in optimized mode to skip children
    const { root, args } = transformWithHoist(`<div/>`)
    expect(root.hoists.length).toBe(0)
    expect(args).toEqual([`"div"`])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist simple element', () => {
    const { root, args } = transformWithHoist(
      `<div><span class="inline">hello</span></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"span"`,
          createObjectMatcher({ class: 'inline' }),
          {
            type: NodeTypes.TEXT,
            content: `hello`
          }
        ]
      }
    ])
    expect(args).toMatchObject([
      `"div"`,
      `null`,
      [
        {
          type: NodeTypes.ELEMENT,
          codegenNode: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `_hoisted_1`
          }
        }
      ]
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist nested static tree', () => {
    const { root, args } = transformWithHoist(
      `<div><p><span/><span/></p></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"p"`,
          `null`,
          [
            { type: NodeTypes.ELEMENT, tag: `span` },
            { type: NodeTypes.ELEMENT, tag: `span` }
          ]
        ]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist nested static tree with comments', () => {
    const { root, args } = transformWithHoist(
      `<div><div><!--comment--></div></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"div"`,
          `null`,
          [{ type: NodeTypes.COMMENT, content: `comment` }]
        ]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist siblings with common non-hoistable parent', () => {
    const { root, args } = transformWithHoist(`<div><span/><div/></div>`)
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      },
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"div"`]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      },
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_2`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist components', () => {
    const { root, args } = transformWithHoist(`<div><Comp/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [`_component_Comp`]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist element with dynamic props', () => {
    const { root, args } = transformWithHoist(`<div><div :id="foo"/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            createObjectMatcher({
              id: `[foo]`
            }),
            `null`,
            genFlagText(PatchFlags.PROPS),
            `["id"]`
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist element with static key', () => {
    const { root, args } = transformWithHoist(`<div><div key="foo"/></div>`)
    expect(root.hoists.length).toBe(1)
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"div"`, createObjectMatcher({ key: 'foo' })]
      }
    ])
    expect(args).toMatchObject([
      `"div"`,
      `null`,
      [
        {
          type: NodeTypes.ELEMENT,
          codegenNode: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `_hoisted_1`
          }
        }
      ]
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist element with dynamic key', () => {
    const { root, args } = transformWithHoist(`<div><div :key="foo"/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            createObjectMatcher({
              key: `[foo]`
            })
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist element with dynamic ref', () => {
    const { root, args } = transformWithHoist(`<div><div :ref="foo"/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            createObjectMatcher({
              ref: `[foo]`
            }),
            `null`,
            genFlagText(PatchFlags.NEED_PATCH)
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with directives', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo" v-foo/></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: WITH_DIRECTIVES,
          arguments: [
            {
              callee: CREATE_VNODE,
              arguments: [
                `"div"`,
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: `_hoisted_1`
                },
                `null`,
                genFlagText(PatchFlags.NEED_PATCH)
              ]
            },
            {
              type: NodeTypes.JS_ARRAY_EXPRESSION
            }
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with dynamic text children', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo">{{ hello }}</div></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            { type: NodeTypes.INTERPOLATION },
            genFlagText(PatchFlags.TEXT)
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with unhoistable children', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo"><Comp/></div></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            [{ type: NodeTypes.ELEMENT, tag: `Comp` }]
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should hoist v-if props/children if static', () => {
    const { root, args } = transformWithHoist(
      `<div><div v-if="ok" id="foo"><span/></div></div>`
    )
    expect(root.hoists).toMatchObject([
      createObjectMatcher({
        key: `[0]`, // key injected by v-if branch
        id: 'foo'
      }),
      {
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      }
    ])
    expect(args[2][0].codegenNode).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          type: NodeTypes.JS_CONDITIONAL_EXPRESSION,
          consequent: {
            // blocks should NOT be hoisted
            callee: CREATE_BLOCK,
            arguments: [
              `"div"`,
              { content: `_hoisted_1` },
              [
                {
                  codegenNode: { content: `_hoisted_2` }
                }
              ]
            ]
          }
        }
      ]
    })
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should hoist v-for children if static', () => {
    const { root, args } = transformWithHoist(
      `<div><div v-for="i in list" id="foo"><span/></div></div>`
    )
    expect(root.hoists).toMatchObject([
      createObjectMatcher({
        id: 'foo'
      }),
      {
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      }
    ])
    const forBlockCodegen = args[2][0].codegenNode
    expect(forBlockCodegen).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          callee: CREATE_BLOCK,
          arguments: [
            FRAGMENT,
            `null`,
            {
              type: NodeTypes.JS_CALL_EXPRESSION,
              callee: RENDER_LIST
            },
            genFlagText(PatchFlags.UNKEYED_FRAGMENT)
          ]
        }
      ]
    })
    const innerBlockCodegen =
      forBlockCodegen.expressions[1].arguments[2].arguments[1].returns
    expect(innerBlockCodegen).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          callee: CREATE_BLOCK,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            [
              {
                codegenNode: { content: `_hoisted_2` }
              }
            ]
          ]
        }
      ]
    })
    expect(generate(root).code).toMatchSnapshot()
  })

  describe('prefixIdentifiers', () => {
    test('hoist nested static tree with static interpolation', () => {
      const { root, args } = transformWithHoist(
        `<div><span>foo {{ 1 }} {{ true }}</span></div>`,
        {
          prefixIdentifiers: true
        }
      )
      expect(root.hoists).toMatchObject([
        {
          type: NodeTypes.JS_CALL_EXPRESSION,
          callee: CREATE_VNODE,
          arguments: [
            `"span"`,
            `null`,
            [
              {
                type: NodeTypes.TEXT,
                content: `foo `
              },
              {
                type: NodeTypes.INTERPOLATION,
                content: {
                  content: `1`,
                  isStatic: false,
                  isConstant: true
                }
              },
              {
                type: NodeTypes.TEXT,
                content: ` `
              },
              {
                type: NodeTypes.INTERPOLATION,
                content: {
                  content: `true`,
                  isStatic: false,
                  isConstant: true
                }
              }
            ]
          ]
        }
      ])
      expect(args).toMatchObject([
        `"div"`,
        `null`,
        [
          {
            type: NodeTypes.ELEMENT,
            codegenNode: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: `_hoisted_1`
            }
          }
        ]
      ])
      expect(generate(root).code).toMatchSnapshot()
    })

    test('hoist nested static tree with static prop value', () => {
      const { root, args } = transformWithHoist(
        `<div><span :foo="0">{{ 1 }}</span></div>`,
        {
          prefixIdentifiers: true
        }
      )

      expect(root.hoists).toMatchObject([
        {
          type: NodeTypes.JS_CALL_EXPRESSION,
          callee: CREATE_VNODE,
          arguments: [
            `"span"`,
            createObjectMatcher({ foo: `[0]` }),
            {
              type: NodeTypes.INTERPOLATION,
              content: {
                content: `1`,
                isStatic: false,
                isConstant: true
              }
            }
          ]
        }
      ])
      expect(args).toMatchObject([
        `"div"`,
        `null`,
        [
          {
            type: NodeTypes.ELEMENT,
            codegenNode: {
              type: NodeTypes.SIMPLE_EXPRESSION,
              content: `_hoisted_1`
            }
          }
        ]
      ])
      expect(generate(root).code).toMatchSnapshot()
    })

    test('hoist class with static object value', () => {
      const { root, args } = transformWithHoist(
        `<div><span :class="{ foo: true }">{{ bar }}</span></div>`,
        {
          prefixIdentifiers: true
        }
      )

      expect(root.hoists).toMatchObject([
        {
          type: NodeTypes.JS_OBJECT_EXPRESSION,
          properties: [
            {
              key: {
                content: `class`,
                isConstant: true,
                isStatic: true
              },
              value: {
                content: `{ foo: true }`,
                isConstant: true,
                isStatic: false
              }
            }
          ]
        }
      ])
      expect(args).toMatchObject([
        `"div"`,
        `null`,
        [
          {
            type: NodeTypes.ELEMENT,
            codegenNode: {
              callee: CREATE_VNODE,
              arguments: [
                `"span"`,
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: `_hoisted_1`
                },
                {
                  type: NodeTypes.INTERPOLATION,
                  content: {
                    content: `_ctx.bar`,
                    isConstant: false,
                    isStatic: false
                  }
                },
                `1 /* TEXT */`
              ]
            }
          }
        ]
      ])
      expect(generate(root).code).toMatchSnapshot()
    })

    test('should NOT hoist expressions that refer scope variables', () => {
      const { root } = transformWithHoist(
        `<div><p v-for="o in list"><span>{{ o }}</span></p></div>`,
        {
          prefixIdentifiers: true
        }
      )

      expect(root.hoists.length).toBe(0)
      expect(generate(root).code).toMatchSnapshot()
    })

    test('should NOT hoist expressions that refer scope variables (2)', () => {
      const { root } = transformWithHoist(
        `<div><p v-for="o in list"><span>{{ o + 'foo' }}</span></p></div>`,
        {
          prefixIdentifiers: true
        }
      )

      expect(root.hoists.length).toBe(0)
      expect(generate(root).code).toMatchSnapshot()
    })

    test('should NOT hoist expressions that refer scope variables (v-slot)', () => {
      const { root } = transformWithHoist(
        `<Comp v-slot="{ foo }">{{ foo }}</Comp>`,
        {
          prefixIdentifiers: true
        }
      )

      expect(root.hoists.length).toBe(0)
      expect(generate(root).code).toMatchSnapshot()
    })

    test('should NOT hoist elements with cached handlers', () => {
      const { root } = transformWithHoist(
        `<div><div><div @click="foo"/></div></div>`,
        {
          prefixIdentifiers: true,
          cacheHandlers: true
        }
      )

      expect(root.cached).toBe(1)
      expect(root.hoists.length).toBe(0)
      expect(
        generate(root, {
          mode: 'module',
          prefixIdentifiers: true
        }).code
      ).toMatchSnapshot()
    })
  })
})
