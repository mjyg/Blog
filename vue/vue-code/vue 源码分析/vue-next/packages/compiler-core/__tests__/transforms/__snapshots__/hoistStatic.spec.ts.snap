// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: hoistStatic transform hoist element with static key 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"div\\", { key: \\"foo\\" })

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist nested static tree 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"p\\", null, [
  _createVNode(\\"span\\"),
  _createVNode(\\"span\\")
])

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist nested static tree with comments 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode
const _createCommentVNode = Vue.createCommentVNode

const _hoisted_1 = _createVNode(\\"div\\", null, [
  _createCommentVNode(\\"comment\\")
])

return function render() {
  with (this) {
    const { createCommentVNode: _createCommentVNode, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist siblings with common non-hoistable parent 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\")
const _hoisted_2 = _createVNode(\\"div\\")

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1,
      _hoisted_2
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist simple element 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\", { class: \\"inline\\" }, \\"hello\\")

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with directives 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { createVNode: _createVNode, withDirectives: _withDirectives, resolveDirective: _resolveDirective, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _directive_foo = _resolveDirective(\\"foo\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _withDirectives(_createVNode(\\"div\\", _hoisted_1, null, 32 /* NEED_PATCH */), [
        [_directive_foo]
      ])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with dynamic text children 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { toString: _toString, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", _hoisted_1, _toString(hello), 1 /* TEXT */)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with unhoistable children 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { resolveComponent: _resolveComponent, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _component_Comp = _resolveComponent(\\"Comp\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", _hoisted_1, [
        _createVNode(_component_Comp)
      ])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers hoist class with static object value 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { class: { foo: true }}

return function render() {
  with (this) {
    const { toString: _toString, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"span\\", _hoisted_1, _toString(_ctx.bar), 1 /* TEXT */)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers hoist nested static tree with static interpolation 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\", null, [
  \\"foo \\",
  _toString(1),
  \\" \\",
  _toString(true)
])

return function render() {
  with (this) {
    const { toString: _toString, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers hoist nested static tree with static prop value 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\", { foo: 0 }, _toString(1))

return function render() {
  with (this) {
    const { toString: _toString, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers should NOT hoist elements with cached handlers 1`] = `
"import { createVNode, createBlock, openBlock } from \\"vue\\"

export default function render() {
  const _ctx = this
  const _cache = _ctx.$cache
  return (openBlock(), createBlock(\\"div\\", null, [
    createVNode(\\"div\\", null, [
      createVNode(\\"div\\", {
        onClick: _cache[1] || (_cache[1] = $event => (_ctx.foo($event)))
      })
    ])
  ]))
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers should NOT hoist expressions that refer scope variables (2) 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { renderList: _renderList, openBlock: _openBlock, createBlock: _createBlock, Fragment: _Fragment, toString: _toString, createVNode: _createVNode } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(false), _createBlock(_Fragment, null, _renderList(_ctx.list, (o) => {
        return (_openBlock(), _createBlock(\\"p\\", null, [
          _createVNode(\\"span\\", null, _toString(o + 'foo'), 1 /* TEXT */)
        ]))
      }), 128 /* UNKEYED_FRAGMENT */))
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers should NOT hoist expressions that refer scope variables (v-slot) 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { toString: _toString, resolveComponent: _resolveComponent, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _component_Comp = _resolveComponent(\\"Comp\\")
    
    return (_openBlock(), _createBlock(_component_Comp, null, {
      default: ({ foo }) => [_toString(_ctx.foo)],
      _compiled: true
    }))
  }
}"
`;

exports[`compiler: hoistStatic transform prefixIdentifiers should NOT hoist expressions that refer scope variables 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { renderList: _renderList, openBlock: _openBlock, createBlock: _createBlock, Fragment: _Fragment, toString: _toString, createVNode: _createVNode } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(false), _createBlock(_Fragment, null, _renderList(_ctx.list, (o) => {
        return (_openBlock(), _createBlock(\\"p\\", null, [
          _createVNode(\\"span\\", null, _toString(o), 1 /* TEXT */)
        ]))
      }), 128 /* UNKEYED_FRAGMENT */))
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist components 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { resolveComponent: _resolveComponent, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _component_Comp = _resolveComponent(\\"Comp\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(_component_Comp)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist element with dynamic key 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", { key: foo })
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist element with dynamic props 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", { id: foo }, null, 8 /* PROPS */, [\\"id\\"])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist element with dynamic ref 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", { ref: foo }, null, 32 /* NEED_PATCH */)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist root node 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\"))
  }
}"
`;

exports[`compiler: hoistStatic transform should hoist v-for children if static 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }
const _hoisted_2 = _createVNode(\\"span\\")

return function render() {
  with (this) {
    const { renderList: _renderList, openBlock: _openBlock, createBlock: _createBlock, Fragment: _Fragment, createVNode: _createVNode } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(false), _createBlock(_Fragment, null, _renderList(list, (i) => {
        return (_openBlock(), _createBlock(\\"div\\", _hoisted_1, [
          _hoisted_2
        ]))
      }), 128 /* UNKEYED_FRAGMENT */))
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should hoist v-if props/children if static 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode
const _createCommentVNode = Vue.createCommentVNode

const _hoisted_1 = {
  key: 0,
  id: \\"foo\\"
}
const _hoisted_2 = _createVNode(\\"span\\")

return function render() {
  with (this) {
    const { openBlock: _openBlock, createVNode: _createVNode, createBlock: _createBlock, createCommentVNode: _createCommentVNode } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(), ok
        ? _createBlock(\\"div\\", _hoisted_1, [
            _hoisted_2
          ])
        : _createCommentVNode(\\"v-if\\", true))
    ]))
  }
}"
`;
